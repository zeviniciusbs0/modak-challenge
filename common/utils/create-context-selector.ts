import { type Context, useContextSelector } from "use-context-selector";

export const createContextSelector =
	<T>(context: Context<T>) =>
	<Selected>(selector: (state: T) => Selected) => {
		return useContextSelector(context, selector);
	};
