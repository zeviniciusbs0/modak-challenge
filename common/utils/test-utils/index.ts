// Re-export everything from React Native Testing Library
export * from "@testing-library/react-native";

// Re-export common Jest functions
export { jest } from "@jest/globals";

import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
	renderHook as originalRenderHook,
	render,
} from "@testing-library/react-native";
import type {
	RenderHookOptions,
	RenderOptions,
} from "@testing-library/react-native";
// Custom render function with providers
import React from "react";
import { TamaguiProvider } from "tamagui";
import { tamaguiConfig } from "../../../tamagui.config";

// Create QueryClient with test-friendly defaults
const createTestQueryClient = () =>
	new QueryClient({
		defaultOptions: {
			queries: {
				retry: false,
				staleTime: Number.POSITIVE_INFINITY,
				gcTime: 0, // Immediate garbage collection
			},
			mutations: {
				retry: false,
			},
		},
	});

// Create a custom render function that includes providers
const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	const queryClient = createTestQueryClient();

	return React.createElement(
		TamaguiProvider,
		{ config: tamaguiConfig, defaultTheme: "light" },
		React.createElement(QueryClientProvider, { client: queryClient }, children),
	);
};

const customRender = (
	ui: React.ReactElement,
	options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: Wrapper, ...options });

const customRenderHook = <Result, Props>(
	hook: (initialProps: Props) => Result,
	options?: Omit<RenderHookOptions<Props>, "wrapper">,
) => originalRenderHook(hook, { wrapper: Wrapper, ...options });

// Re-export everything, overriding render and adding renderHook
export { customRender as render, customRenderHook as renderHook };

// Test utilities
export const createMockNavigation = () => ({
	navigate: jest.fn(),
	goBack: jest.fn(),
	dispatch: jest.fn(),
	reset: jest.fn(),
	canGoBack: jest.fn(() => true),
	isFocused: jest.fn(() => true),
});

export const createMockRoute = (params = {}) => ({
	key: "test-route",
	name: "TestScreen",
	params,
});

// Helper to wait for async operations
export const waitForNextTick = () =>
	new Promise((resolve) => setTimeout(resolve, 0));

// Mock data generators
export const createMockProduct = (overrides = {}) => ({
	id: 1,
	title: "Test Product",
	description: "Test Description",
	price: 99.99,
	category: "test-category",
	thumbnail: "https://test.com/image.jpg",
	images: ["https://test.com/image.jpg"],
	rating: 4.5,
	stock: 10,
	tags: ["test"],
	brand: "Test Brand",
	sku: "TEST-SKU",
	weight: 1,
	dimensions: { width: 10, height: 10, depth: 10 },
	warrantyInformation: "Test warranty",
	shippingInformation: "Test shipping",
	availabilityStatus: "In Stock",
	reviews: [],
	returnPolicy: "Test return",
	minimumOrderQuantity: 1,
	meta: {
		createdAt: "2024-01-01",
		updatedAt: "2024-01-01",
		barcode: "test-barcode",
		qrCode: "test-qr",
	},
	discountPercentage: 0,
	...overrides,
});

export const createMockCategory = (overrides = {}) => ({
	slug: "test-category",
	name: "Test Category",
	url: "https://dummyjson.com/products/category/test-category",
	...overrides,
});
