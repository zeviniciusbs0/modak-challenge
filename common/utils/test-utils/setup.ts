jest.mock("@/common/config/notifications", () => ({
	useConfigureNotificationsHandler: jest.fn(() => ({
		expoPushToken: "ExponentPushToken[test]",
		notification: null,
	})),
	sendLocalNotification: jest.fn(),
}));

jest.mock("axios", () => ({
	create: jest.fn(() => ({
		get: jest.fn(),
		post: jest.fn(),
		put: jest.fn(),
		delete: jest.fn(),
	})),
	get: jest.fn(),
	post: jest.fn(),
	put: jest.fn(),
	delete: jest.fn(),
}));

global.console = {
	...console,
	warn: jest.fn(),
	error: jest.fn(),
};

afterEach(() => {
	jest.runOnlyPendingTimers();
	jest.useRealTimers();
});

afterAll(() => {
	return new Promise((resolve) => setTimeout(resolve, 100));
});
