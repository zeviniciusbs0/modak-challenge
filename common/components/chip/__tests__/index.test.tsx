import { fireEvent, jest, render } from "@/common/utils/test-utils";
import { Chip } from "../index";

describe("Chip", () => {
	it("should render the chip correctly", () => {
		const { toJSON } = render(<Chip>Chip Test</Chip>);

		expect(toJSON()).toMatchSnapshot();
	});

	describe("interaction priority logic", () => {
		it("should prioritize onRemove over onPress when both provided", () => {
			const mockPress = jest.fn();
			const mockRemove = jest.fn();

			const { getByText } = render(
				<Chip onPress={mockPress} onRemove={mockRemove}>
					Priority Test
				</Chip>,
			);

			const chip = getByText("Priority Test");
			fireEvent.press(chip);

			expect(mockRemove).toHaveBeenCalledTimes(1);
			expect(mockPress).not.toHaveBeenCalled();
		});

		it("should use onPress when onRemove not provided", () => {
			const mockPress = jest.fn();

			const { getByText } = render(<Chip onPress={mockPress}>Press Test</Chip>);

			const chip = getByText("Press Test");
			fireEvent.press(chip);

			expect(mockPress).toHaveBeenCalledTimes(1);
		});
	});
});
