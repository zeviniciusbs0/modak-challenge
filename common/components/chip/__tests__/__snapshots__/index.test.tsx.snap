// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip should render the chip correctly 1`] = `
<View
  focusVisibleStyle={
    {
      "backgroundColor": "transparent",
      "borderBottomColor": "hsla(146, 38%, 69%, 1)",
      "borderLeftColor": "hsla(146, 38%, 69%, 1)",
      "borderRightColor": "hsla(146, 38%, 69%, 1)",
      "borderTopColor": "hsla(146, 38%, 69%, 1)",
    }
  }
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderBottomColor": "hsla(146, 38%, 69%, 1)",
      "borderBottomLeftRadius": 7,
      "borderBottomRightRadius": 7,
      "borderBottomWidth": 2,
      "borderLeftColor": "hsla(146, 38%, 69%, 1)",
      "borderLeftWidth": 2,
      "borderRightColor": "hsla(146, 38%, 69%, 1)",
      "borderRightWidth": 2,
      "borderStyle": "solid",
      "borderTopColor": "hsla(146, 38%, 69%, 1)",
      "borderTopLeftRadius": 7,
      "borderTopRightRadius": 7,
      "borderTopWidth": 2,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 36,
      "justifyContent": "center",
      "paddingLeft": 13,
      "paddingRight": 13,
    }
  }
>
  <Text
    lineBreakMode="clip"
    numberOfLines={1}
    style={
      {
        "color": "hsla(155, 41%, 14%, 1)",
        "flexGrow": 0,
        "flexShrink": 1,
        "fontFamily": "System",
        "fontSize": 13,
        "fontWeight": "300",
        "letterSpacing": 0,
        "lineHeight": 23,
        "userSelect": "none",
      }
    }
    suppressHighlighting={true}
  >
    Chip Test
  </Text>
</View>
`;
